contracts/Delay.sol

```diff
@@ -80,90 +136,198 @@ contract Delay is AragonApp, IForwarder {
-
-        DelayedScript memory delayedScript = delayedScripts[_delayedScriptId];
-        delete delayedScripts[_delayedScriptId];
+    function execute(uint256 _delayedScriptId) external nonReentrant {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        require(_canExecute(delayedScript), ERROR_CAN_NOT_EXECUTE);
 
         runScript(delayedScript.evmCallScript, new bytes(0), new address[](0));
-        emit ExecutedScript(_delayedScriptId);
+
+        if (delayedScript.disputableStatus != DisputableStatus.Closed) {
+            _closeAgreementAction(delayedScript.actionId);
+        }
+
+        emit ExecutedScript(_delayedScriptId, delayedScript.actionId);
+        _deleteDelayedScript(_delayedScriptId);
+    }
+
+    /**
+    * @notice Close agreement action with ID `_delayedScriptId`
+    * @param _delayedScriptId The ID of the script to close the agreement action of
+    */
+    function closeAgreementAction(uint256 _delayedScriptId) external {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+
+        require(_canExecute(delayedScript), ERROR_CAN_NOT_EXECUTE);
+        require(delayedScript.disputableStatus != DisputableStatus.Closed, ERROR_ACTION_CLOSED);
+
+        _closeAgreementAction(delayedScript.actionId);
+        emit AgreementActionClosed(_delayedScriptId, delayedScript.actionId);
+
+        delayedScript.disputableStatus = DisputableStatus.Closed;
     }
 
     /**
     * @notice Return whether a script with ID #`_scriptId` can be executed
-    * @param _scriptId The ID of the script to execute
+    * @param _delayedScriptId The ID of the script to execute
     */
-    function canExecute(uint256 _scriptId) public view returns (bool) {
-        bool withinExecutionWindow = getTimestamp64() > delayedScripts[_scriptId].executionTime;
-        bool isUnpaused = !_isExecutionPaused(_scriptId);
+    function canExecute(uint256 _delayedScriptId) public view returns (bool) {
+        DelayedScript storage delayedScript = delayedScripts[_delayedScriptId];
+        return _canExecute(delayedScript);
+    }
 
-        return withinExecutionWindow && isUnpaused;
+    /**
+    * @dev IForwarder interface conformance
+    */
+    function isForwarder() external pure returns (bool) {
+        return true;
     }
 
+    /**
+    * @dev IForwarder interface conformance
+    */
     function canForward(address _sender, bytes) public view returns (bool) {
         return canPerform(_sender, DELAY_EXECUTION_ROLE, arr());
     }
 
     /**
     * @notice Delays execution for `@transformTime(self.executionDelay(): uint)`
+    * @dev IForwarder interface conformance
     * @param _evmCallScript The script that can be executed after a delay
     */
     function forward(bytes _evmCallScript) public {
         require(canForward(msg.sender, _evmCallScript), ERROR_CAN_NOT_FORWARD);
-        _delayExecution(_evmCallScript);
+        _delayExecution(new bytes(0), _evmCallScript);
```

Comment:
Friendly reminder we will need to update this with the [new forwarding interface](https://github.com/aragon/aragonOS/pull/583/files#diff-bc2d5c7bc9b406f0a55ea357c99410c9)

Reply:
Ah yeah, I saw in the Voting app, will update when it's available.